language/tools/move-cli/src/main.rs

```diff
@@ -529,8 +567,54 @@ fn explain_type_error(
     println!("Execution failed with type error when binding type arguments to type parameters")
 }
 
+fn explain_publish_error(
+    error: VMError,
+    state: &OnDiskStateView,
+    module: &CompiledModule,
+) -> Result<()> {
+    use StatusCode::*;
+
+    let module_id = module.self_id();
+    match error.into_vm_status() {
+        VMStatus::Error(DUPLICATE_MODULE_NAME) => {
+            println!(
+                "Module {} exists already. Re-run without --no-republish to publish anyway.",
+                module_id
+            );
+        }
+        VMStatus::Error(BACKWARD_INCOMPATIBLE_MODULE_UPDATE) => {
+            if let Ok(old_m) = state.get_compiled_module(&module_id) {
+                let old_api = Module::new(&old_m);
+                let new_api = Module::new(module);
+                let compat = Compatibility::check(&old_api, &new_api);
+                if !compat.is_fully_compatible() {
```

Comment:
Should we assert this instead of using an `if`? I think it should always be true if we get a `BACKWARD_INCOMPATIBLE_MODULE_UPDATE`

Reply:
Right, we should assert here. I'll make a change.